### PR TITLE
linux-libre: init

### PR DESCRIPTION
**Pull request by alyssais**
###### Motivation for this change

GNU Linux-libre is a linux kernel distribution that has all non-free binary blobs stripped. The project distributes both deblobbed kernel sources, and scripts for each kernel version that perform the deblobbing. Here, I have used the scripts rather than the pre-deblobbed sources, because this allows an arbitrary kernel to be passed in and deblobbed with no extra effort, which seems more in the spirit of Nix and allows deblobbing to be composed with other kernel customisations.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

With the small fix to openafs included in this PR, `linuxPackages-libre` and `linuxPackages_latest-libre` can build all kernel modules that the corresponding non-deblobbed kernels can build. (This does not mean that `nox-review` will pass, because some kernel modules are currently broken on non-deblobbed Linux 4.18 anyway. See #45410.) The fix to openafs does not break any existing kernels where it was previously able to build. I have booted and used a deblobbed kernel successfully, with the exception of some of my laptop's hardware not working, which is to be expected with a deblobbed kernel.

### CODE REVIEW

File: pkgs/top-level/all-packages.nix [hunk 1]
```diff
@@ -14210,6 +14210,12 @@ with pkgs;
   linuxPackages_hardkernel_latest = linuxPackages_hardkernel_4_14;
   linux_hardkernel_latest = linuxPackages_hardkernel_latest.kernel;
 
+  # GNU Linux-libre kernels
+  linuxPackages-libre = recurseIntoAttrs (linuxPackagesFor linux-libre);
+  linux-libre = callPackage ../os-specific/linux/kernel/linux-libre.nix {};
+  linuxPackages_latest-libre = recurseIntoAttrs (linuxPackagesFor linux_latest-libre);
+  linux_latest-libre = linux-libre.override { linux = linux_latest; };
```
Comment (Mic92) [inline pkgs/top-level/all-packages.nix]: Do we really need both latest and stable?

File: pkgs/os-specific/linux/kernel/linux-libre.nix [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+}:
+
+let
+  majorMinor = lib.versions.majorMinor linux.modDirVersion;
+
+  major = lib.versions.major linux.modDirVersion;
+  minor = lib.versions.minor linux.modDirVersion;
+  patch = lib.versions.patch linux.modDirVersion;
+
+in linux.override {
```
Comment (Mic92) [inline pkgs/os-specific/linux/kernel/linux-libre.nix]: For things like kernels, I would like to see a maintainer.

File: pkgs/top-level/all-packages.nix (hunk 1 above)
Comment (alyssais) [reply pkgs/top-level/all-packages.nix]: I did this for consistency with other kernels – hardened and xen_dom0 both have latest and stable. I can definitely see both having an audience.

**Comment by xeji** [comment]
Pushed a trivial fix so we can actually build a system with this kernel. LGTM now.

**Comment by xeji** [comment]
selection_002

**Review by xeji — APPROVED** [review]

**Comment by Mic92** [comment]
I suppose non-gpl third-party modules should be blacklisted for this kernel? https://hydra.nixos.org/eval/1476119?filter=linuxPackages-libre&compare=1476064&full=#tabs-new

**Pull request merged**